PR 3/5 - Refactor(favorite) implement favorite controller (#842)

AdriaVH opened the pull request:
# PR  3/5 Summary - Task 842: Implement Favorite Controller

This PR corresponds to **[Task 842](https://tree.taiga.io/project/luisvi-ita-challenges/task/842)** and introduces the REST controller responsible for exposing the new Favorites API endpoint.

It follows the work completed in:
- **[Task 840](https://tree.taiga.io/project/luisvi-ita-challenges/task/840)**
- **[Task 841](https://tree.taiga.io/project/luisvi-ita-challenges/task/841)**

It is the **third step** towards isolating and refactoring the GET Favorites functionality away from `User` and into the new dedicated Favorites collection.

------------------------------------------------------------------------

## Key changes

### 1. New Favorite Controller

A new controller has been implemented under:

com.itachallenge.user.controller.userinteraction.favorite.FavoriteController

This controller exposes the new endpoint:

    GET /itachallenge/api/v1/user/users/{userId}/favorites

It retrieves all challenges marked as favorite by the specified user and delegates the operation to the new `FavoriteService` introduced in Task 841.

### 2. Controller Behavior

-   Returns **Mono\<Set`<UUID>`\>**, preserving the existing contract expected by the front-end
-   Handles validation and errors through the existing `UserGlobalExceptionHandler`
-   Enriched with OpenAPI annotations (`@Operation`, `@ApiResponse`, etc.) for improved API documentation
-   Logs key events for observability

### 3. Maintains Backwards Compatibility

⚠️ Although `FavoriteResponseDto` exists, it is **not yet used** in order to preserve the current front-end expectations:
➡️ The endpoint returns an **Array of UUIDs**
➡️ No breaking changes introduced

------------------------------------------------------------------------

## ✅ Created Files

### Controller

-   `FavoriteController.java`
    -  getUserFavorites(@PathVariable String userId)

### Tests

-   `FavoriteControllerTest.java`
    -   Covers successful response, invalid UUID, and user not found
        scenarios
    -   Uses `WebTestClient` and the global exception handler

------------------------------------------------------------------------

## 🔄 Modified Files

- `UserController`
      - <getUserFavorites(@PathVariable String userId)
None beyond what is required for wiring dependencies and tests.
All existing behaviors remain unchanged.

------------------------------------------------------------------------

## 💡 Notes

-   This PR exposes the **first public API** consuming the new Favorites collection.
-   The new collection and services are now fully functional, but GET Favorites still returns UUID arrays to maintain backward compatibility.

-------------------------------------------------------------------------

## Technical debt :

A technical debt has been created in order to change the path of the FavoriteController once all endpoints have been migrated : [issue 863](https://tree.taiga.io/project/luisvi-ita-challenges/issue/863)

------------------------------------------------------------------------

# ✅ Impact for API Consumers

No changes to existing API responses or formats.
Consumers continue receiving a simple Array of UUIDs.

------------------------------------------------------------------------

# ✅ Changelog

Changelog no need of modification.

------------------------------------------------------------------------

# ✅ PR Author Self-Check

-   [x] New FavoriteController created under the expected package
-   [x] Path `/itachallenge/api/v1/user/users/{userId}/favorites` exposed
-   [x] Delegated logic to FavoriteService
-   [x] OpenAPI documentation applied
-   [x] Controller covered by unit tests with WebTestClient
-   [x] No breaking changes for API consumers
-   [x] Code follows project conventions and Sonar rules
-   [x] All tests pass locally and remotely

itsAidenDev committed:
Redistribution of functions to new controller + added tests

itsAidenDev committed:
Redistribution of functions to new controller + added tests

AdriaVH committed:
Deleted gitkeep, not needed anymore

AdriaVH committed:
- Added getUserFavorites from User scope to FavoriteController

AdriaVH committed:
- Added getUserFavorites from User scope to FavoriteController

AdriaVH committed:
- Added getUserFavorites from User scope to FavoriteController

AdriaVH committed:
- Added getUserFavorites from User scope to FavoriteController

AdriaVH committed:
deleted gitkeep

AdriaVH committed:
added RestController and mapping

AdriaVH committed:
added RestController and mapping

AdriaVH committed:
added RestController and mapping

AdriaVH committed:
added RestController and mapping

AdriaVH committed:
added Tests from User

AdriaVH committed:
moved getUserFavorites from UserController test ot FavoriteControlerTest

AdriaVH committed:
moved getUserFavorites from UserController test ot FavoriteControlerTest

zohra-b committed:
added UserRepository in FavoriteServiceImpl + quitted all add - delete tests accordind to how it is in 841

AdriaVH committed:
refactor UserController to remove FavoriteService dependency and delegate favorite operations to UserService

AdriaVH committed:
refactor UserController to remove FavoriteService dependency and delegate favorite operations to UserService

AdriaVH committed:
refactor UserServiceImpl to expose addToFavorites and deleteFromFavorites methods

AdriaVH committed:
refactor UserControllerTest to remove getUserFavorites tests and streamline bookmark tests

AdriaVH committed:
refactor UserControllerTest to remove getUserFavorites tests and streamline bookmark tests

Pablo2203 reviewed (changes requested):
Hello Adria!

Great job! I have only one doubt regarding your PR, it seems like the PR dragged changes made on previous PRs of the chronology. I´m saying this because I´m not sure if this PR you touched the class UserService that it would not be in your scope.

There are also lot of phantom changes but i don´t think it was your fault.

Also in the documentation you wrote that Changelog would be modified (correct me if I´m wrong) and I do not see those changes.

Check the test to pass. the Sonar and I think that´s all.

AdriaVH committed:
refactor UserController to remove getUserFavorites method and clean up code

AdriaVH commented:
Great job! I have only one doubt regarding your PR, it seems like the PR dragged changes made on previous PRs of the chronology. I´m saying this because I´m not sure if this PR you touched the class UserService that it would not be in your scope.

**You are absolutely right, we did mess up a bit and got carried away at the beginning and we did Services and Controller in these branch at first. But then did it properly and applied and modified services in 841 branch and only do the Controler and its tests here as per documented in PR**

There are also lot of phantom changes but i don´t think it was your fault.

**Should be clenaned now**

Also in the documentation you wrote that Changelog would be modified (correct me if I´m wrong) and I do not see those changes.

**Also right again, my bad, i have modified that now from the PR documentation, as no Changelog modification**

Check the test to pass. the Sonar and I think that´s all. 

**Tests passed same with Sonar, thanks and loads of love**

@Pablo2203

FlavioKde reviewed (approved):
Good work — UserService aligned with the pull request and changes correctly implemented.

- The new controller is well-placed and respects the original path.

- The logic was decoupled from the UserService and delegated to the FavoriteService.

- Full compatibility with the frontend is maintained (same format and status).

- The tests were migrated and cover all expected scenarios.


Great refactoring!

AdriaVH committed:
Remove favorite management methods from UserService

Removed methods for adding and deleting favorites in UserService.

AdriaVH committed:
Change methods to private in UserServiceImpl

Changed visibility of addToFavorites and deleteFromFavorites methods from public to private.

AdriaVH reviewed (commented):
- methods from UserServiceImpl back to private
-  deleted the methods from UserService

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/controller/userinteraction/favorite/FavoriteController.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.UUID;
+
+@RestController
+@RequestMapping("/itachallenge/api/v1/user")
```
⚠️ Currently, the FavoriteController maintains the same @RequestMapping("/itachallenge/api/v1/user") as the UserController, to avoid breaking the front end in this sprint.

However, this is not a best practice, as it mixes user and favorites resources under the same path.

🧩 Technical debt:
Once the migration of all favorites endpoints is complete, we should update the base route to... (You suggest one)

This way, the module will be properly isolated and the resource will be semantically consistent with the REST API.

📅 This change should be made in a later sprint, when the front-end team has the time to adapt the routes and associated tests.

In the PR description, please note that a technical debt has been created and attach the URL of the request. Thank you.

ivilarop replied:
Technical debt created: https://tree.taiga.io/project/luisvi-ita-challenges/issue/863

AdriaVH committed:
Update CHANGELOG for version 3.0.4 ControlerFavorite

Added FavoriteController and refactored UserController to improve modularity and testability.

ivilarop commented on CHANGELOG.md:
```diff
@@ -3,6 +3,16 @@
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+### [itachallenge-user-3.0.4-RELEASE] - 2025-11-13
```
change to correct version

AdriaVH replied:
changed

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/controller/userinteraction/favorite/FavoriteControllerTest.java:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    @Test
+    void getUser_WhenUserExists_Returns200() {
```
These tests are not for favorites; they should be in the userController tests.

AdriaVH replied:
okay

ivilarop replied:
ok thks!

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/controller/userinteraction/favorite/FavoriteControllerTest.java:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    @Test
+    void getUser_WhenUserNotExists_Returns404() {
```
These tests are not for favorites; they should be in the userController tests.

AdriaVH replied:
done, thank you

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/controller/userinteraction/favorite/FavoriteControllerTest.java:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    @Test
+    void getUser_WhenServiceReturnsError_Returns500() {
```
These tests are not for favorites; they should be in the userController tests.

AdriaVH replied:
done

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/controller/userinteraction/favorite/FavoriteControllerTest.java:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    @Test
+    void testEndpoint_ShouldReturnHelloMessage() {
```
This test is for userController. Shouldn't a new test endpoint be created in favoriteController and tested there?

AdriaVH replied:
the test of endpoint is unnecessary i will delete it

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/controller/userinteraction/favorite/FavoriteControllerTest.java:
```diff
@@ -0,0 +1,165 @@
+    @BeforeEach
+    void setUp() {
+        mocks = MockitoAnnotations.openMocks(this);
+        webTestClient = WebTestClient.bindToController(userController, new FavoriteController(favoriteService))
```
That userController in bindToController(...) shouldn't be there if the test is for FavoriteController.

Furthermore, with `@WebFluxTest(FavoriteController.class)`, Spring already provides a ready-made WebTestClient and the controller wired. You don't need `setUp()` to build anything.

You would only use `@BeforeEach` if you want to initialize data (e.g., `userId = UUID.randomUUID()`), not to create the WebTestClient.

AdriaVH replied:
Okay, Before Each deleted and@WebFluxTest implement, thanks

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/controller/userinteraction/favorite/FavoriteControllerTest.java:
```diff
@@ -0,0 +1,165 @@
+    @InjectMocks
+    private UserController userController;
+
+    private WebTestClient webTestClient;
```
inject with @Autowired

ivilarop replied:
implemented

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/controller/userinteraction/favorite/FavoriteControllerTest.java:
```diff
@@ -0,0 +1,165 @@
+class FavoriteControllerTest {
+
+    @Mock
+    private UserService userService;
```
Is it necessary?

AdriaVH replied:
nope, thanks

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/controller/userinteraction/favorite/FavoriteControllerTest.java:
```diff
@@ -0,0 +1,165 @@
+    private FavoriteService favoriteService;
+
+    @Mock
+    private IUserSolutionService userSolutionService;
```
Is it necessary?

AdriaVH replied:
not really, cheers

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/controller/userinteraction/favorite/FavoriteControllerTest.java:
```diff
@@ -0,0 +1,165 @@
+    private IUserSolutionService userSolutionService;
+
+    @InjectMocks
+    private UserController userController;
```
Is it necessary?

AdriaVH replied:
wow, yes, ofc, sorry

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/controller/userinteraction/favorite/FavoriteControllerTest.java:
```diff
@@ -0,0 +1,165 @@
+
+    private WebTestClient webTestClient;
+
+    private AutoCloseable mocks;
```
Is it necessary?

AdriaVH replied:
absolutely right, thanks

ivilarop commented on itachallenge-user/src/test/java/com/itachallenge/user/controller/userinteraction/favorite/FavoriteControllerTest.java:
```diff
@@ -0,0 +1,165 @@
+
+import static org.mockito.Mockito.*;
+
+class FavoriteControllerTest {
```
@WebFluxTest(FavoriteController.class)
@Import(UserGlobalExceptionHandler.class)

to avoid the beforeEach

AdriaVH replied:
implemented, thank you

ivilarop replied:
ok thks!

ivilarop commented on itachallenge-user/src/main/java/com/itachallenge/user/controller/userinteraction/favorite/FavoriteController.java:
```diff
@@ -0,0 +1,55 @@
+            }
+    )
+    @GetMapping("/users/{userId}/favorites")
+    public Mono<ResponseEntity<Set<UUID>>> getUserFavorites(@PathVariable String userId) {
```
Do I need to modify anything in ITA-Challenges.postman_collection.json or add the new endpoint for get Favorites?

AdriaVH replied:
No, still going thru User for now

ivilarop replied:
ok thks!

AdriaVH committed:
modified to correct version

AdriaVH committed:
modifications from Ivan's review applied

AdriaVH committed:
Merge branch 'develop' into refactor/842-Create-FavoriteController
